### PR TITLE
feat(#2760): implement D-1b multi-file GeneralCoder

### LINKED ISSUE #2760: D-1b: Multi-file GeneralCoder (≤5 files)
## D-1b: Multi-file GeneralCoder (≤5 files)

**Parent EPIC**: #2759 (EPIC D: Autonomous Coder Agent Family)
**Stage**: 1b - Multi-file Extension
**Priority**: High
**前置條件**: D-1a (#3344) 已完成

---

## 目標

擴展 SimpleCoder 為支援多檔案編輯的 GeneralCoder，能夠處理跨檔案的代碼變更。

---

## 與 D-1a (SimpleCoder) 的差異

| 項目 | D-1a SimpleCoder | D-1b GeneralCoder |
|------|------------------|-------------------|
| 檔案數 | 單一檔案 | 多檔案 (≤5) |
| Import 理解 | 無 | 有 |
| Schema | CoderOutput (單檔案) | MultiFileCoderOutput (多檔案) |
| Feature Flag | 無 | ENABLE_GENERAL_CODER |

---

### 1. Multi-file Schema Extension
- 擴展 `CoderOutput` 支援 `patches: List[FilePatch]`
- 或改用 unified diff 格式
- 更新 drift detection tests

### 2. Multi-file Context Manager
- 能同時讀取多個檔案的內容
- 理解檔案間的依賴關係
- 限制：最多 5 個檔案

### 3. Apply/Commit Semantics
- 支援多檔案原子提交
- 或定義明確的 rollback 策略

### 4. Guardrails Scaling
- Per-file syntax checks
- Path allowlist / repo-root enforcement
- "No new files / no deletes unless allowed"

---

## 驗收條件

- [ ] 能處理多檔案編輯（≤5 個檔案）
- [ ] 能理解檔案間的 import 關係
- [ ] 能被 Flow Controller v3 正確路由
- [ ] 有單元測試覆蓋主要場景
- [ ] Feature flag 控制 (ENABLE_GENERAL_CODER)
- [ ] #3288 (whitespace-only patches) 已解決

---

## 關聯 Issues

- D-1a: #3344 (SimpleCoder MVP - 已完成)
- D-2: #2761 (Senior Coder Logic - 依賴此 Issue)
- D-4: #2764 (Self-Correction Loop - 依賴此 Issue)
- #3288: SimpleCoder whitespace-only patches

---

## 歷史記錄

- 2025-12-30: 從原 D-1 拆分為 D-1b，D-1a (#3344) 已完成並關閉
- 原 D-1 的單檔案能力已在 D-1a 中實現

---

Link to Devin run: https://app.devin.ai/sessions/9bab8bd3f6904a1db27deca6fd525ac4
Requested by: Ryan Chen (@RC918)

### PR DESCRIPTION
**Pull request by RC918**
# feat(#2760): implement D-1b multi-file GeneralCoder

## Summary

Implements EPIC D Phase 1b: Multi-file auto-fix capability for the autonomous coder agent family. This adds a new `GeneralCoder` agent that can edit up to 5 files atomically, with import relationship understanding and per-file syntax validation.

**Key changes:**
- New `GeneralCoder` class extending `BaseAgent` with `MultiFileCoderOutput` schema
- New `commit_files()` function using GitHub Git Data API for atomic multi-file commits
- `ENABLE_GENERAL_CODER` feature flag (default: False for safe rollout)
- Integration into `fixer_node` with fallback chain: GeneralCoder → SimpleCoder → AutoFixer
- Python import parsing via `ast` module for understanding file relationships

**Guardrails implemented:**
- Max 5 files per operation
- Per-file Python syntax validation
- Path traversal protection (blocks `..`, absolute paths, uses `os.path.normpath()`)
- Whitespace-only patch rejection (Issue #3288)

## Updates since last revision

- **Addressed gemini-code-assist feedback**: Moved `import os` to top of `general_coder.py` (PEP 8 compliance)
- **Consistent path validation**: Added `os.path.normpath()` to `commit_files()` in `github_api.py` to match GeneralCoder's validation logic
- **Improved path validation**: Both GeneralCoder and commit_files now use the same path validation pattern: normalize with `os.path.normpath()` before checking for `..`
- **Created follow-up issue**: #3350 for future import resolution improvements with posixpath (low priority, modernization-p3)
- **Addressed MorningAI Reviewer feedback**: Responded to 3 comments - rejected schema version bump (parallel schemas by design), created follow-up for import resolution, implemented path validation hardening

## Review & Testing Checklist for Human

- [ ] **Verify atomic commit semantics in `commit_files()`** - The function uses a 6-step GitHub Git Data API flow (get_git_ref → get_git_commit → create_git_tree → create_git_commit → ref.edit). If step 5 succeeds but step 6 fails, the commit exists but branch isn't updated. Review lines 584-812 in `github_api.py`.

- [ ] **Review path validation security** - Path traversal protection now uses `os.path.normpath()` before checking for `..` in both locations. Verify this handles all edge cases (lines 517-530 in `general_coder.py`, lines 644-666 in `github_api.py`).

- [ ] **Test feature flag integration** - Set `ENABLE_GENERAL_CODER=True` in staging and verify the fallback chain works correctly (GeneralCoder → SimpleCoder → AutoFixer).

- [ ] **Verify LLM response parsing robustness** - The `generate_multi_file_fix()` method parses JSON from LLM. Test with malformed responses to ensure graceful degradation.

**Recommended test plan:**
1. Deploy to staging with `ENABLE_GENERAL_CODER=False` (default) - verify no behavior change
2. Enable flag and trigger a multi-file review scenario
3. Verify atomic commit creates single commit with all file changes
4. Test fallback by having GeneralCoder skip (complex issue) and verify SimpleCoder takes over

### Notes

- All 107 tests pass (35 GeneralCoder + 72 commit_file)
- Feature flag is OFF by default - no production impact until explicitly enabled
- The `_attempt_general_coder_fix` function has high cyclomatic complexity (34) - this is flagged by lint but acceptable for this integration point
- Schema design: `MultiFileCoderOutput` is a **parallel schema** to `CoderOutput`, not a modification. Both use `schema_version=1` as they are v1 of their respective schema families.

Closes #2760

---
Link to Devin run: https://app.devin.ai/sessions/9bab8bd3f6904a1db27deca6fd525ac4
Requested by: Ryan Chen (ryan2939z@gmail.com) / @RC918

### CODE REVIEW

File: handoff/20250928/40_App/orchestrator/coder/general_coder.py [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    patches: List[FilePatch] = field(default_factory=list)
+    files_affected: int = 0
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Convert to dictionary for JSON serialization.
+
+        Returns the Final MultiFileCoderOutput Schema with schema_version.
+        """
+        result: Dict[str, Any] = {
+            "schema_version": CODER_OUTPUT_SCHEMA_VERSION,
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/coder/general_coder.py]: IMPACT: The MultiFileCoderOutput schema introduces new fields (schema_version, syntax_valid, files_affected) without incrementing the schema version. This breaks backward compatibility for consumers expecting the old format from SimpleCoder. SUGGESTED FIX: Increment CODER_OUTPUT_SCHEMA_VERSION to 2 in simple_coder.py and document the breaking changes.

**Review by RC918 — COMMENTED** [review]
## MorningAI Code Review

File: handoff/20250928/40_App/orchestrator/coder/general_coder.py [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+                if not patch_content.strip():
+                    logger.warning(
+                        f"[GENERAL_CODER_WHITESPACE_ABORT] {file_path}: whitespace-only content"
+                    )
+                    return MultiFileCoderOutput.create_skipped(
+                        f"Whitespace-only patch for {file_path}"
+                    )
+
+                # Path validation (no traversal)
+                if ".." in file_path or file_path.startswith("/"):
```
[the remark this answers is not in the document]
Comment (RC918) [reply handoff/20250928/40_App/orchestrator/coder/general_coder.py]: **評估結果：部分採納**

有效的安全考量，但需要澄清：
1. URL encoding (`%2e%2e`) 不適用 - 路徑來自 LLM 輸出，不是 user input
2. Windows paths 不適用 - Linux 部署環境
3. Symlinks 不適用 - 透過 GitHub API 建立新檔案，不是 follow local symlinks

**採納的改進**：將使用 `os.path.normpath()` 在檢查 `..` 之前正規化路徑，以處理 `foo/../bar` 這類 edge case。這與 `commit_files()` 的驗證邏輯保持一致。
